packages/sitecore-jss-nextjs/src/edge/personalize-middleware.test.ts

```diff
@@ -0,0 +1,958 @@
+  const userAgentStub = sinon.stub(nextjs, 'userAgent').returns({ ua: 'ua' } as any);
+  const debugSpy = spy(debug, 'personalize');
+  const validateDebugLog = (message, ...params) =>
+    expect(debugSpy.args.find((log) => log[0] === message)).to.deep.equal([message, ...params]);
+
+  const createMiddleware = (
+    props: { [key: string]: unknown; cdpConfig?: any; edgeConfig?: any } = {}
+  ) =>
+    new PersonalizeMiddleware({
+      ...props,
+      cdpConfig: {
+        clientKey: 'cdp-client-key',
+        endpoint: 'http://cdp-endpoint',
+        pointOfSale: 'cdp-pos',
+        ...(props?.cdpConfig || {}),
+      },
+      edgeConfig: {
+        apiKey: 'edge-api-key',
+        endpoint: 'http://edge-endpoint/api/graph/edge',
+        siteName: 'nextjs-app',
+        ...(props?.edgeConfig || {}),
+      },
+    });
+
+  const id = 'item-id';
+  const version = '1';
+  const variantIds = ['variant-1', 'variant-2'];
+  const browserId = 'browser-id';
+  const contentId = `${id}_en_${version}`.toLowerCase();
+  const personalizeQueryResult = {
+    layout: {
+      item: {
+        id,
+        version,
+        personalization: {
+          variantIds,
+        },
+      },
+    },
+  };
+
+  const referrer = 'http://localhost:3000';
+  const experienceParams: ExperienceParams = {
+    geo: {
+      city: 'geo-city',
+      country: 'geo-country',
+      latitude: 'geo-latitude',
+      longitude: 'geo-longitude',
+      region: 'geo-region',
+    },
+    referrer,
+    ua: 'ua',
+    utm: {
+      campaign: 'utm_campaign',
+      content: null,
+      medium: null,
+      source: null,
+    },
+  };
+
+  const createRequest = (props: any = {}) => {
+    const req = {
+      ...props,
+      nextUrl: {
+        pathname: '/styleguide',
+        locale: 'en',
+        searchParams: {
+          get(key) {
+            return {
+              utm_campaign: 'utm_campaign',
+              utm_content: null,
+              utm_medium: null,
+              utm_source: null,
+            }[key];
+          },
+        },
+        clone() {
+          return Object.assign({}, req.nextUrl);
+        },
+        ...props?.nextUrl,
+      },
+      cookies: {
+        get(cookieName: string) {
+          const cookies = { 'bid_cdp-client-key': 'browser-id', ...props.cookieValues };
+
+          return cookies[cookieName];
+        },
+        ...props?.cookies,
+      },
+      referrer,
+      geo: props?.geo === null ? null : props?.geo || experienceParams.geo,
+    } as NextRequest;
+
+    return req;
+  };
+
+  const createResponse = (props: any = {}) => {
+    const res = {
+      cookies: {
+        set(key, value) {
+          res.cookies[key] = value;
+        },
+      },
+      headers: {
+        set(key, value) {
+          res.headers[key] = value;
+        },
+      },
+      ...props,
+    } as NextResponse;
+
+    return res;
+  };
+
+  const mockPersonalizeDataRequest = (data: any = personalizeQueryResult) =>
```

Comment:
I think we are getting too deep on the mocking here. We shouldn't have to worry about the underlying fetch requests of the `GraphQLPersonalizeService` and `CdpService`. Instead, we should be mocking those classes directly and performing all stubs/expectations against their APIs.

I believe we could stub the private variables (`personalizeService`/`cdpService`), but likely the more proper way would be to add these dependencies to the constructor (i.e. add these to the `PersonalizeMiddlewareConfig`) so we can inject our mocks.

Reply:
Good catch! Done